src/agent/solana/oracle.rs

```diff
@@ -687,62 +694,70 @@ mod subscriber {
                 wss_url,
                 rpc_timeout,
                 commitment,
-                account_key,
+                program_key,
                 updates_tx,
                 logger,
             }
         }
 
         pub async fn run(&self) {
-            match self.start_shadow().await {
-                Ok(mut shadow_rx) => self.forward_updates(&mut shadow_rx).await,
-                Err(err) => {
-                    error!(self.logger, "{}", err);
-                    debug!(self.logger, "error context"; "context" => format!("{:?}", err));
-                }
-            }
-        }
-
-        async fn forward_updates(&self, shadow_rx: &mut broadcast::Receiver<(Pubkey, Account)>) {
             loop {
-                if let Err(err) = self.forward_update(shadow_rx).await {
+                let current_time = Instant::now();
+                if let Err(ref err) = self.start().await {
                     error!(self.logger, "{}", err);
                     debug!(self.logger, "error context"; "context" => format!("{:?}", err));
+                    if current_time.elapsed() < Duration::from_secs(30) {
+                        tracing::error!(
+                            "Subscriber restarting too quickly. Sleeping for 1 second."
+                        );
+                        tokio::time::sleep(Duration::from_secs(1)).await;
+                    }
                 }
             }
         }
 
-        async fn forward_update(
-            &self,
-            shadow_rx: &mut broadcast::Receiver<(Pubkey, Account)>,
-        ) -> Result<()> {
-            self.updates_tx
-                .send(shadow_rx.recv().await?)
+        pub async fn start(&self) -> Result<()> {
+            debug!(self.logger, "subscribed to program account updates"; "program_key" => self.program_key.to_string());
```

Comment:
Shouldn't this be logged at the end of pubsub client setup?

Reply:
It was there in the past but good point. Moved it below.